hail/src/main/scala/is/hail/backend/Backend.scala

```diff
@@ -1,16 +1,120 @@
+case class ImportFamPayload(path: String, quant_pheno: Boolean, delimiter: String, missing: String)
+case class ExecutePayload(ir: String, stream_codec: String, timed: Boolean)
+
+object BackendServer {
+  def apply(backend: Backend) = new BackendServer(backend)
+}
+
+class BackendServer(backend: Backend) {
+  // 0 => let the OS pick an available port
+  private val httpServer = HttpServer.create(new InetSocketAddress(0), 10)
+  private val handler = new BackendHttpHandler(backend)
+
+  def port = httpServer.getAddress.getPort
+
+  def start(): Unit = {
+    httpServer.createContext("/", handler)
+    httpServer.setExecutor(null)
+    httpServer.start()
+  }
+
+  def stop(): Unit = {
+    httpServer.stop(10)
+  }
+}
+
+class BackendHttpHandler(backend: Backend) extends HttpHandler {
+  def handle(exchange: HttpExchange): Unit = {
+    implicit val formats: Formats = DefaultFormats
+
+    try {
+      val body = using(exchange.getRequestBody)(JsonMethods.parse(_))
+      if (exchange.getRequestURI.getPath == "/execute") {
+          val config = body.extract[ExecutePayload]
+          backend.execute(config.ir, config.timed) { (ctx, res, timings) =>
+            exchange.getResponseHeaders().add("X-Hail-Timings", timings)
+            res match {
+              case Left(_) => exchange.sendResponseHeaders(200, -1L)
+              case Right((t, off)) =>
+                exchange.sendResponseHeaders(200, 0L)  // 0 => an arbitrarily long response body
+                using(exchange.getResponseBody()) { os =>
+                  backend.encodeToOutputStream(ctx, t, off, config.stream_codec, os)
+                }
+            }
+          }
+          return
+      }
+      val response: Array[Byte] = exchange.getRequestURI.getPath match {
+        case "/value/type" => backend.valueType(body.extract[IRTypePayload].ir)
+        case "/table/type" => backend.tableType(body.extract[IRTypePayload].ir)
+        case "/matrixtable/type" => backend.matrixTableType(body.extract[IRTypePayload].ir)
+        case "/blockmatrix/type" => backend.blockMatrixType(body.extract[IRTypePayload].ir)
+        case "/references/load" => backend.loadReferencesFromDataset(body.extract[LoadReferencesFromDatasetPayload].path)
+        case "/references/from_fasta" =>
+          val config = body.extract[FromFASTAFilePayload]
+          backend.fromFASTAFile(config.name, config.fasta_file, config.index_file,
+            config.x_contigs, config.y_contigs, config.mt_contigs, config.par)
+        case "/vcf/metadata/parse" => backend.parseVCFMetadata(body.extract[ParseVCFMetadataPayload].path)
+        case "/fam/import" =>
+          val config = body.extract[ImportFamPayload]
+          backend.importFam(config.path, config.quant_pheno, config.delimiter, config.missing)
+      }
+
+      exchange.sendResponseHeaders(200, response.length)
+      using(exchange.getResponseBody())(_.write(response))
+    } catch {
+      case t: Throwable =>
+        error(t.getMessage)
```

Comment:
Is this a log statement? Can we use `log.error` instead?

Reply:
This was debugging cruft, the error will get logged in python anyway. Removed